File: C++/Problems/Problems/problem_1010.cpp
#include "problem_1010.h"

bool problem_1010::test() {
    std::vector time  { 30, 20, 150, 100, 40 };

    int expected = 3;

    int result = numPairsDivisibleBy60(time);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_1010::numPairsDivisibleBy60(std::vector<int> &time) {
    int n = time.size();
    std::vector rests(60, 0);
    int result = 0;

    for (int i = 0; i < n; i++)
    {
        int rest = time[i] % 60;
        int expected = (60 - rest) % 60;
        result += rests[expected];
        rests[rest]++;
    }

    return result;
}

File: Python/Problems/Problems/problem1010.py
from typing import List

from .problem import Problem


class Problem1010(Problem):
    def test(self) -> bool:
        time = [30, 20, 150, 100, 40]

        expected = 3

        result = self.numPairsDivisibleBy60(time)

        print(result)

        return result == expected

    def numPairsDivisibleBy60(self, time: List[int]) -> int:
        n = len(time)
        rests = [0] * 60
        result = 0

        for i in range(n):
            rest = time[i] % 60
            expected = (60 - rest) % 60
            result += rests[expected]
            rests[rest] += 1

        return result